daemon/server/router/system/info_response_test.go

```diff
@@ -31,3 +34,58 @@ func TestLegacyFields(t *testing.T) {
 		t.Errorf("legacy fields should contain %s: %s", expected, string(data))
 	}
 }
+
+// TestMarshalRegistryConfigLegacyFields verifies extra fields in the registry config
+// field in the info response are serialized if they are not empty.
+// This is used for backwards compatibility for API versions < 1.49.
+func TestMarshalRegistryConfigLegacyFields(t *testing.T) {
+	expected := map[string]string{
+		"AllowNondistributableArtifactsCIDRs":     `"AllowNondistributableArtifactsCIDRs": null`,
+		"AllowNondistributableArtifactsHostnames": `"AllowNondistributableArtifactsHostnames": null`,
+	}
+
+	tests := []struct {
+		name   string
+		info   *infoResponse
+		assert func(t *testing.T, data []byte, err error)
+	}{
+		{
+			name: "without legacy fields",
+			info: &infoResponse{
+				Info: &system.Info{},
+			},
+			assert: func(t *testing.T, data []byte, err error) {
+				assert.NilError(t, err)
+				for k, v := range expected {
+					assert.Check(t, !strings.Contains(string(data), v), fmt.Sprintf("Legacy field %q should be omitted in output: %s", k, string(data)))
+				}
+			},
+		},
+		{
+			name: "with legacy fields",
+			info: &infoResponse{
+				Info: &system.Info{
+					RegistryConfig: &registry.ServiceConfig{
+						ExtraFields: map[string]any{
+							"AllowNondistributableArtifactsCIDRs":     json.RawMessage(nil),
+							"AllowNondistributableArtifactsHostnames": json.RawMessage(nil),
+						},
+					},
+				},
+			},
+			assert: func(t *testing.T, data []byte, err error) {
+				assert.NilError(t, err)
+				for k, v := range expected {
+					assert.Check(t, strings.Contains(string(data), v), fmt.Sprintf("Expected legacy field %q to be included in the output: %s", k, string(data)))
```

Comment:
I think `assert.Check(t, is.Contains(string(data), v))` may also work for strings, and print the output when failing

(writing from my phone so can't verify what it looks like)